### DISCUSSION TITLE
Other ways to handle unrestorable buffers

### DISCUSSION THREAD

**Discussion by hab25** [alphapapa/activities.el#85] [Ideas]
My use-case: I use activities as a substitute for `desktop-mode` that improves the experience by  providing the added benefit of working for many non-file-backed buffers. 

I do this with code very similar to the following in at the top level of my `init.el`:

```emacs-lisp
(run-with-idle-timer
  5
  nil
  (lambda () 
	(let ((my/default-activity-name "my/default"))
        ;; side note: please make `#'activities-named` be autoloaded; as can be seen here, it is an appropriate entry point and would allow me to remove this `#'require`. 
	    (require 'activities)
	    (activities-resume
	      (or
	        (activities-named my/default-activity-name)
	        (activities-define my/default-activity-name))))))
```

However, in doing so, I often hit https://github.com/alphapapa/activities.el/blob/b63ccc58fc3b5b3e6ed5b74bf84bd4eff9bb1660/activities.el#L796 

Such a customization would allow me to avoid saving such buffers that cannot be restored.

**Comment by alphapapa** [comment]
Hello,

Adding such a predicate might be useful, but please help me understand your use case better.  What are you trying to do with that idle timer?

**Comment by hab25** [comment]
> but please help me understand your use case better

Gladly.

> What are you trying to do with that idle timer?

I am trying to avoid waiting for `activities` to load the default activity on startup, instead deferring to when I am next idle. This is useful when starting up a fresh Emacs instance to test out new changes to my `init.el`, where I want to test the new behaviour as fast as possible and am not so interested in the buffers I was previously working with.

**Comment by alphapapa** [comment]
Thanks, but I'm still not sure I completely understand, because `activities` doesn't load an activity on Emacs startup.  Also, the function in the timer also calls `activities-define` sometimes.  Would you elaborate, please?

**Comment by hab25** [comment]
> Would you elaborate, please?

Sure.

---

> I'm still not sure I completely understand, because activities doesn't load an activity on Emacs startup.

Correct. I mistakenly understood your question

> What are you trying to do with that idle timer?

as "Why use a timer?" instead of "What is going on inside that `run-with-idle-timer` form?".

So, to add and clarify to my previous comment, if I were to not use a timer, i.e. have instead the following at the top-level of my `init.el`,

```
(let ((my/default-activity-name "my/default"))
        ;; side note: please make `#'activities-named` be autoloaded; as can be seen here, it is an appropriate entry point and would allow me to remove this `#'require`. 
	    (require 'activities)
	    (activities-resume
	      (or
	        (activities-named my/default-activity-name)
	        (activities-define my/default-activity-name))))
```

I would have to suffer what the timer avoids, that is, 

> waiting for `activities` to load the default activity on startup

---

> Also, the function in the timer also calls activities-define sometimes. Would you elaborate, please?

If `(activities-named my/default-activity-name)` returns nil, an activity named `my/default-activity-name` doesn't exist yet; I use the default value of `activities-always-persist`,   which is `t`; this case means we are in first-ever run of this code on this computer. Therefore, now run `(activities-define my/default-activity-name)` to create it.

Please feel free to ask further.

**Comment by alphapapa** [comment]
Forgive me, but you seem to have translated the Lisp into English, and I can already read code.  I'm asking about what you're trying to accomplish with that code.

**Comment by hab25** [comment]
> what you're trying to accomplish with that code.

I am trying to restore, on Emacs startup, the buffers that were live right before the last Emacs session ended.

**Comment by alphapapa** [comment]
Ok, thanks.  Generally it's best to lead with the root problem you're trying to solve rather than an incidental one several steps down the line--or, at least, to mention both.  (cf. "the XY problem")

Now, what kinds of buffers are you getting those errors for?  As I've mentioned elsewhere, the best solution is to make those buffers bookmarkable so they won't have such errors, and it usually doesn't take much code to do so.

**Comment by hab25** [comment]
> Now, what kinds of buffers are you getting those errors for?

Buffers that don't have bookmark support; n.b. the link in the opening post.

Here's a recent example:

> Activities was unable to get a buffer named: *Memory-Profiler-Report 2024-04-22 00:47:08*
>
> It is likely that this buffer's major mode does not support the `bookmark' system, so it can't be restored properly.  Please ask the major mode's maintainer to add bookmark support.
>
> If this is not the case, please report this error to the `activities' maintainer.
>
> In the meantime, it's recommended to not use buffers of this major mode in an activity's layout; or you may simply ignore this error and use the other buffers in the activity.

In the previous session, this buffer had been created by `profiler-report` after having invoked `profiler-start` and choosing "cpu+mem".

> Generally it's best to lead with the root problem you're trying to solve rather than an incidental one several steps down the line--or, at least, to mention both. (cf. "the XY problem")

Agreed. I thought it was made clear by "I use activities as a substitute for desktop-mode" and the name "my/default". In hindsight, I see that `desktop-mode` is flexible enough that my usage couldn't be unambiguously implied.

**Comment by hab25** [comment]
Here's another example:

> Activities was unable to get a buffer named: *eat*
>
> It is likely that this buffer's major mode does not support the `bookmark' system, so it can't be restored properly.  Please ask the major mode's maintainer to add bookmark support.
>
> If this is not the case, please report this error to the `activities' maintainer.
> 
> In the meantime, it's recommended to not use buffers of this major mode in an activity's layout; or you may simply ignore this error and use the other buffers in the activity.

which comes from https://codeberg.org/akib/emacs-eat/src/branch/master/eat.el. Note that this is a terminal emulator; adding (intuitive) bookmark support there probably means, at least, getting and restoring the current working directory from the shell if `eat` is running one; such an integration may not be trivial to implement depending on which shell is being used. 

Therefore,

>  Please ask the major mode's maintainer to add bookmark support.

in this case may not result in bookmark support getting added in a near future.

**Comment by alphapapa** [comment]
Thanks for the examples.

Here are some thoughts:

1. A save-buffers predicate seems like an obvious thing to have, but it would also introduce another question: if a window's buffer does not pass the predicate, what should be done instead?  Should a previously displayed buffer in that window that does pass the predicate be saved instead?  That might not be useful.  Or, what if no such previously displayed buffer passes the predicate?  Should an error buffer be shown, as is currently the case, or should the window be deleted?  Deleting the window would likely be a bad idea, as it would corrupt the window configuration from what was set by the user.
2. Instead, what if the behavior of what to do in the case of an unrestoreable buffer were configurable?  e.g. rather than a buffer showing the error message as currently happens, what if the user could specify to show the scratch buffer, or something else?  This might be especially useful for your specific case, i.e. to only override the default behavior when restoring your last-used activity at startup.
3.  Another possibly useful option would be to save previous window configurations in the activity, ala `winner-mode` and `tab-bar-history-mode`, so that if some buffers weren't restorable, you could go back in the frame's/tab's window configuration history.

Well, those are the best ideas I have.  There is no perfect solution, other than having every major mode support bookmarks.  Certain ones, like profiler reports, are probably impractical and not very useful (although I could imagine a bookmark for the profiler report buffer saving and restoring the contents--that wouldn't seem too difficult, and it could be useful if a debugging session were interrupted).

Also, regarding terminal buffers: Having worked with this functionality for a while in `burly` and `bufler` before `activities`, having terminal buffers support bookmarks is an oft-requested feature.  There's probably no perfect solution, but it probably wouldn't be too hard to save the current working directory and open a new terminal to it when the bookmark is restored.  Of all the terminal emulator libraries, I'm guessing `eat` might be the one most suited to having such functionality (although Eshell is another obvious candidate), so I'd encourage you to discuss the idea with its maintainer.

**Comment by hab25** [comment]
To me, 2) and 3) would be the most useful. They could be combined by passing the history described in 3) as a parameter to a user-supplied function used to configure 2).

I agree about eat being well suited for bookmark support, and that the perfect solution would be for every major mode to implement bookmark support, but AFAIK there is no evidence for believing that this perfection will be reality in a near future, if ever.

**Reply by alphapapa** [reply]
Ok, since this is now a discussion about ideas, I'm converting it to a discussion.  Issues can be opened for more specific ideas (e.g. the numbered ones) as needed to track their implementation.

No, there will likely never be perfection, since we can't control what major modes do, and they can come from anywhere.  So the question for us is how best to handle such situations, and how far it's reasonable to go to try to smooth it over.

FWIW, I'm glad if `activities` is helpful to you in this situation (restoring the last state when you exited Emacs), but note that it's not specifically intended for that; one of the reasons the resetting feature exists is that there will always be situations where an activity's last state includes unrestorable buffers.  But the way you're using a "default activity" means that the resetting doesn't apply, and so the unrestorable buffers will sometimes be an issue.

Anyway, it would likely be good to implement one or more of the ideas here, but I don't plan to work on it myself anytime soon.  Patches welcome, but note that it would probably be good to discuss an implementation idea before coding it, so as to avoid going down paths that I would decline to merge.  

Thanks.

**Comment by capfredf** [comment]
I agree that it is hard to have a general solution, but I think we can probably have some recipes or examples to show how unrestorable buffers from different major modes can be handled.  I always have some temporary buffers which are not intended to be restored. For example, in the `agda2-mode`, there is always a shared singleton window displaying a buffer called `Agda Information` where type information are provided with respect to certain positions. It is shared in that if I have multiple agda files open, I need to reload `Agda Information` whenever I switch buffers.  So in my case, the best approach would be to kill the window and then kill the buffer before the activity is saved.

**Reply by alphapapa** [reply]
> So in my case, the best approach would be to kill the window and then kill the buffer before the activity is saved.

That probably would not be good, because when `activities-mode` is enabled (which is recommended), activities are saved after 5 seconds of idle time.  So that would mean killing that window and buffer while you're using it.

The current "solution" is to either ignore the unrestorable buffer, kill it, or revert the activity to its default state.

Probably the best solution would be to improve that `Agda Information` buffer's implementation to so it could be restored.  It would probably only take a few lines of code and then it would "just work".
